Reset parts data

BeeMargarida opened the pull request:
| - | - |
| --- | --- |
| Issue | When changing the model/brand/version of the `ripe-configurator` and `ripe-image`, an error is thrown due to the `partsData` being outdated with the data of the previous build. <br>![2020-11-03_16-49](https://user-images.githubusercontent.com/25725586/98018438-72336d80-1df8-11eb-8016-3cf4455754f6.png)|
| Dependencies | -- |
| Decisions | When the brand/model/version changes, the `partsData` is updated with the value in the props `parts`, meaning that if the user wants to change everything, including parts, the setup will allow. This solves the outdated data. The parts were removed in the `ripe-image` story to allow the user to experiment this and not break. <br> The setPartsRipe function was fixed. <br> A tiny typo in the pickers was also fixed (i know it is not in the topic of this PR, but it is a tiny tiny fix). |
| Animated GIF | ![Peek 2020-11-03 17-20](https://user-images.githubusercontent.com/25725586/98018833-ecfc8880-1df8-11eb-9fc4-f65c07ef8b4b.gif)|

BeeMargarida committed:
Override partsData when changing ripe brand/model/version

The partsData variable is updated when a change to the parts is made. When switching models, this variable must be updated with the props value so that it does not use values of the previous model

BeeMargarida committed:
Fix

BeeMargarida committed:
Fix typo

BeeMargarida committed:
Removed parts from story

Causes errors when testing the changing of model

BeeMargarida committed:
Fix function call

gcandal commented on vue/components/organisms/ripe-configurator/ripe-configurator.vue:
```diff
@@ -313,6 +313,7 @@ export const RipeConfigurator = {
         },
         configProps: {
             handler: async function(value) {
+                this.partsData = this.parts;
```
Why isn't this updated by the `parts` watcher?

BeeMargarida replied:
It is updated by the `parts` watcher, however, when a new brand/model is given, this runs first.  When `sdk's .config()` is called, the `partsData` has the parts of the old model, resulting in errors by the `sdk`.

gcandal replied:
Can you find out if it's possible to have control over the order by which the watchers fire? It's not a great practice and we'll end up with this fix anyway, is just so that we understand how is that defined (if it's deterministic).

BeeMargarida replied:
So, I was wrong. The flow of events is this:
1. When the model/brand changes, the watcher on the `parts` runs first and the watcher on the `configProps` is next. This order changes it the watcher for the `configProps` is moved to be positioned above the `parts` watcher.
2. By leaving the order as it was, the error in the console happens because the `setPartsRipe` was being called before the `configRipe`, with the old model defaultConfig but the correct parts for the new model https://github.com/ripe-tech/ripe-sdk/blob/f2f47516129cfe937388bef19c9baa4deeb6c5ea/src/js/base/ripe.js#L1515. This resulted in an error by the sdk but eventually the configuration would be made correct.

To fix this and leave the order as is, a check must be made in the `parts` watcher, and not calling the `setPartsRipe` if the brand/model/version `props` differ from the ones in the current config of the sdk, since the call to change (including the setting of new parts) will be made by the `configRipe` that is called next.

gcandal replied:
Maybe we should be changing the order then. WDYT?

BeeMargarida replied:
I tried it, the problem remains. The `setPartsRipe` happens in the middle of the `setConfig`, resulting in using the outdated configuration to set the parts, which is later updated. The change is applied all the same, but the error still happens and appears in the console.

gcandal commented on vue/components/organisms/ripe-image/ripe-image.stories.js:
```diff
@@ -17,26 +17,6 @@ storiesOf("Organisms", module)
                 type: Number,
                 default: number("Version", 52)
             },
-            parts: {
```
Why?

BeeMargarida replied:
Just to allow testing the changing of model/brand without breaking the sdk, since the parts fed would be from the old one.

gcandal replied:
Is this still a problem? I would expect that I could pass (brand, model, parts) then just change (brand, model) and hope that the parts update themselves.

BeeMargarida committed:
Preventing parts update if configuration will change

BeeMargarida committed:
Pretty

BeeMargarida committed:
Apply changes to the configurator

BeeMargarida committed:
Spacing

gcandal reviewed (changes requested):
Please try explicitly using a flag.

BeeMargarida committed:
Using configurating flag

This flag marks when the ripe-sdk is being configured, not allowing other operations to be made to it.

gcandal commented:
@BeeMargarida is the diff correct? It seems it has lot of extra stuff

BeeMargarida commented:
> @BeeMargarida is the diff correct? It seems it has lot of extra stuff

I tried a rebase 🤦 gonna revert

BeeMargarida committed:
Fixed conflicts

gcandal commented on vue/components/organisms/ripe-configurator/ripe-configurator.vue:
```diff
@@ -228,11 +233,30 @@ export const RipeConfigurator = {
+
+                try {
+                    this.partsData = this.parts;
+                    if (this.config) await this.configRipe();
```
Move this config to an early exit (`if (!this.config) return;`)

gcandal commented on vue/components/organisms/ripe-configurator/ripe-configurator.vue:
```diff
@@ -228,11 +233,30 @@ export const RipeConfigurator = {
 
                 this.partsData = value;
+
+                if (this.configuring) return;
```
Treat this as a pre-condition and make an early exit

BeeMargarida commented:
@gcandal Regarding the removal of `parts` from the `ripe-image` story, it is necessary. Since the `configRipe` function updates the `partsData` with the `parts` variable before making the request to the sdk, if a user changes the brand and model in the storybook, the parts will not be updated with the change, resulting in an error.
I suppose that when using this, the client will be listening to the event `update:parts` from the component and update the given parts accordingly, or when changing model, resets the parts or changes it to the ones he wants.

BeeMargarida committed:
Fixes

BeeMargarida committed:
Revert tiny change

BeeMargarida committed:
Revert tiny change

BeeMargarida committed:
Fix for updating model/brand/version and not changing parts prop

When updating the brand/model/version and maintain the same outdated parts, it is important to check if this is the case in order to reset the partsData variable.

BeeMargarida commented:
@gcandal The fix includes this bulky piece of code, but I believe there is no workaround.
```
// if the model and parts changed but the parts
// stayed the same from the previous model, the
// parts must be cleaned
const resetParts =
          (value.brand !== previous.brand ||
            value.model !== previous.model ||
             value.version !== previous.version) &&
           previous.parts && value.parts &&
            this.equalParts(value.parts, previous.parts);
this.partsData = resetParts ? null : value.parts;
```

If the model/brand/version is changed but the client maintains the same parts, the `partsData` is cleaned.

BeeMargarida committed:
Merge branch 'master' into ms/reset-parts-data

BeeMargarida committed:
Tiny fix

gcandal commented:
Closing as we will have to add this same functionality to https://github.com/ripe-tech/ripe-sdk-components-vue/pull/43 in a different form.
